### ISSUE TITLE
Foreign DispatchQueue type

### ISSUE THREAD

**Issue by bendk** [mozilla/uniffi-rs#1734]
A common pattern we use in [application-services](https://github.com/mozilla/application-services/) is to run SQLite queries using a foreign-managed task queue (`Dispatchers.IO` on Kotlin, `DispatchQueue` on Swift, etc).  Could UniFFI support this as a first-class type?

This would require:
  - Moving it across the FFI (only supporting Foreign -> Rust is probably fine)
  - Adding way to execute Rust synchronous code using the foreign DispatchQueue in two modes:
     - async mode where you await the result
     - fire-and-forget mode where you just want the code to run and don't care about the result.

**Comment by bendk** [comment]
I think `BackgroundQueue` is a nicer name than `DispatchQueue`, since the feature is really about running tasks in a background thread.  I feel like there could be a better name though, open to any suggestions.

**Comment by mhammond** [comment]
"background thread" implies "low priority thread" to me, which I don't think is the actual intent. The actual use-cases here tend to be more about thread blocking than thread priority. I was going to suggest something like `TaskQueue`, but depending on the content the `Task` part of that might itself carry baggage. `WorkQueue`? But I don't think it matters that much though.

**Comment by bendk** [comment]
You're right, it really is about blocking more than anything.  `tokio` separates their threads into "worker threads" and "blocking threads".  What about `BlockingQueue` or `BlockingTaskQueue`?

**Comment by mhammond** [comment]
using the same terminology as tokio is certainly appealing!
